### DISCUSSION TITLE
libcurl put data binary upload zip file

### DISCUSSION THREAD

**Discussion by tagopinath** [curl/curl#12708] [Q&A]
### I did this

Hello,

Previously i tried with curl command values, which sometimes gives issue on retrieving the return values. so i started using libcurl. now i am trying to get one function which works fine in command, but not in libcurl.

curl  -X PUT --data-binary @"somezipfile.zip" "uploadurl"

with command, the above statement works fine.

i am trying to do the same through libcurl, but getting CURLE_URL_MALFORMAT error.

here is the code i am trying

~~~c

static size_t read_callback(char *dest, size_t size, size_t nmemb, void *userp)
{
	struct WriteThis *wt = (struct WriteThis *)userp;
	size_t buffer_size = size*nmemb;

	if (wt->sizeleft) 
	{
		/* copy as much as possible from the source to the destination */
		size_t copy_this_much = wt->sizeleft;
		if (copy_this_much > buffer_size)
			copy_this_much = buffer_size;
		memcpy(dest, wt->readptr, copy_this_much);

		wt->readptr += copy_this_much;
		wt->sizeleft -= copy_this_much;
		return copy_this_much; /* we copied this many bytes */
	}

	return 0; /* no more data left to deliver */
}

std::string url_encode(const std::string& decoded)
{
	const auto encoded_value = curl_easy_escape(nullptr, decoded.c_str(), static_cast<int>(decoded.length()));
	std::string result(encoded_value);
	curl_free(encoded_value);
	return result;
}

void fnUploadZipFile(string zipfname, string uploadurl)
{
	CURL *curlReq;
	CURLcode res;
	struct stat file_info;
	curl_off_t speed_upload, total_time;
	FILE *fd;

	fd = fopen(zipfname.c_str(), "rb"); /* open file to upload */
	if (!fd)
		return; /* cannot continue */

				  /* to get the file size */
	if (stat(zipfname.c_str(), &file_info) != 0)
		return; /* cannot continue */

	curl_global_init(CURL_GLOBAL_ALL);//

	curlReq =  curl_easy_init();
	if (curlReq)
	{
		curl_easy_setopt(curlReq, CURLOPT_READFUNCTION, read_callback);

		/* tell it to "upload" to the URL */
		res = curl_easy_setopt(curlReq, CURLOPT_UPLOAD, 1L);
		
		/* upload to this place */
		string enurl = url_encode(uploadurl);
		res = curl_easy_setopt(curlReq, CURLOPT_URL, enurl.c_str());
		/* set where to read from (on Windows you need to use READFUNCTION too) */

		res = curl_easy_setopt(curlReq, CURLOPT_READDATA, fd);

		/* and give the size of the upload (optional) */
		res = curl_easy_setopt(curlReq, CURLOPT_INFILESIZE_LARGE, (curl_off_t)file_info.st_size);

		res = curl_easy_perform(curlReq);
		/* Check for errors */
		if (res != CURLE_OK) 
		{
			string err = curl_easy_strerror(res);
			fprintf(stderr, "curl_easy_perform() failed: %s\n",
				curl_easy_strerror(res));
		}
		else {
			/* now extract transfer info */
			curl_easy_getinfo(curlReq, CURLINFO_SPEED_UPLOAD_T, &speed_upload);
			curl_easy_getinfo(curlReq, CURLINFO_TOTAL_TIME_T, &total_time);

			fprintf(stderr, "Speed: %lu bytes/sec during %lu.%06lu seconds\n",
				(unsigned long)speed_upload,
				(unsigned long)(total_time / 1000000),
				(unsigned long)(total_time % 1000000));
		}
		/* always cleanup */
		curl_easy_cleanup(curlReq);
	}
	fclose(fd);
}
~~~
what may be the issue? 

Regards,
Gopi.

### I expected the following

need to get some return values

### curl/libcurl version

libcurl 8.5.0

### operating system

windows 10

**Comment by bagder** [comment]
I you take a look at the URL you tell libcurl to work on I think you will notice the problem. You seem to URL encode the entire thing, which I figure is highly unlikely to be a good idea.

**Reply by tagopinath** [reply]
i tried by replacing only space, ?, # and @, then also i am getting the same error.

**Reply by bagder** [reply]
And the URL looks exactly how you want it to look? Then look for leading or trailing garbage/whitespace.

**Reply by jay** [reply]
do not encode the whole url. encode only the filename if you need to. use verbose mode `curl -V` and [`CURLOPT_VERBOSE`](https://curl.se/libcurl/c/CURLOPT_VERBOSE.html) and compare.

**Comment by jay** [comment]
> curl -X PUT --data-binary @"somezipfile.zip" "uploadurl"
> 
> with command, the above statement works fine.

The correct way to do that is with [`curl -T`](https://curl.se/docs/manpage.html#-T) like `curl -T somezipfile.zip https://hostname/directory/filename` or `curl -T somezipfile.zip https://hostname/directory/`. In the latter case there's no specific filename so curl uses the input filename `somezipfile.zip`(eg file is actually uploaded to `https://hostname/directory/somezipfile.zip`.

**Reply by tagopinath** [reply]
command statement is working fine. I need the same functionality in libcurl cpp sdk.
